pkg/expr/transform.go

```diff
@@ -189,9 +189,15 @@ func (s *Service) queryData(ctx context.Context, req *backend.QueryDataRequest)
 		if err != nil {
 			return nil, err
 		}
+
+		calculatedInterval, err := interval.GetIntervalFrom(getDsInfo.Result, sj, time.Second)
```

Comment:
This is the minimum interval. But it might mean the actual interval, depending on who was initiating the request. If request originating from ngalert, t_his is the minimum interval_. If the request originating from HTTP API and the `IntervalMS` is set, it might mean _use this as the interval_.

As an example, for Elasticsearch we use the time range and the min interval to calculate the actual interval:
https://github.com/grafana/grafana/blob/2b22ab6eaf01a1e8e9d825b1cc6a680e806fc424/pkg/tsdb/elasticsearch/time_series_query.go#L72

It's very confusing, but discussed some of this with Gabor a while ago.

Reply:
Yes I think I am confused :-/ 

But I feel like I am calling the "shared" function for this, so if I am calling this function correctly(?), then would additional changes to make this work go into the data source or the `GetIntervalFrom` function?

Reply:
yup, confusing :-) 

as @marefr mentioned, yes, this function is (mostly) meant to get you the minInterval (which is basically, get it from the json-thing (for dashboard/explore queries), or get it from the datasource, or use the min-value sent in).

i wonder, is this place ONLY called from new-alert? or is this place called also from dashboard-queries "old alerts" etc?

if it is only used from new-alert, then, my understanding is that in such case there is no way to "send in" the min-interval, so what will happen here is that datasource-min-interval will be used, or the specified min-value ( `time.Second` ). 

in general, i wonder who should calculate the "final" interval-value: the alerting system or the datasource-plugin.

the thing is, when you run a query from the browser (dashboard or explore-mode), there is some generic code (not datasource-specific) that calculates the final interval value based on: datasource-min-interval,panel-min-interval,time-range, max-data-points (see https://github.com/grafana/grafana/blob/main/public/app/features/query/state/PanelQueryRunner.ts#L231-L232 for the dashboard-case).

so when for example, the flux datasource gets a query it can just take `DataSubQuery.IntervalMS` and send it to the database, no calculation needed. 

so if we want to achieve the same with new-alerting, it also should do the whole calculation based on those factors. 

alternatively, we could say no, the datasource has to do all this, and let's say we set `IntervalMS` to zero or `-1` and the datasource has to do the calculation.

sorry, i don't have a solution, only problems 😄